File: ports/go/src/gui/kernel/bcsinputarbitrator.go
package kernel

import (
	"bcs/core/global"
	"sync"
)

// BcsInputOwner maps the conceptual multi-user network ID to an input source
type BcsInputOwner struct {
	ID   string
	Name string
}

func NewBcsInputOwner(id, name string) *BcsInputOwner {
	return &BcsInputOwner{ID: id, Name: name}
}

// BcsInputArbitrator handles multi-user event streams and focus delegation natively
type BcsInputArbitrator struct {
	mu           sync.RWMutex
	activeOwners map[string]*BcsInputOwner
	FocusChanged global.BcsSignal
}

func NewBcsInputArbitrator() *BcsInputArbitrator {
	return &BcsInputArbitrator{
		activeOwners: make(map[string]*BcsInputOwner),
	}
}

// RegisterOwner adds a new network or local user to the arbitrator's pool
func (a *BcsInputArbitrator) RegisterOwner(owner *BcsInputOwner) {
	a.mu.Lock()
	defer a.mu.Unlock()
	a.activeOwners[owner.ID] = owner
}

// RequestFocus attempts to route the target focus token to the specified owner
func (a *BcsInputArbitrator) RequestFocus(ownerID string, targetWidget *BcsObject) bool {
	a.mu.RLock()
	owner, exists := a.activeOwners[ownerID]
	a.mu.RUnlock()

	if !exists || targetWidget == nil {
		return false
	}

	// Emit signal broadcasting focus shift logic
	a.FocusChanged.Emit(owner)
	return true
}

File: ports/rust/src/gui/kernel/bcsinputarbitrator.rs
use std::collections::HashMap;
use std::sync::{Arc, RwLock};
use std::rc::Rc;
use std::cell::RefCell;
use crate::core::global::bcssignal::BcsSignal;
use crate::core::kernel::bcs_event::BcsObject;

// BcsInputOwner maps the conceptual multi-user network ID to an input source
#[derive(Clone)]
pub struct BcsInputOwner {
    pub id: String,
    pub name: String,
}

impl BcsInputOwner {
    pub fn new(id: &str, name: &str) -> Self {
        BcsInputOwner {
            id: id.to_string(),
            name: name.to_string(),
        }
    }
}

// BcsInputArbitrator handles multi-user event streams and focus delegation natively
pub struct BcsInputArbitrator {
    active_owners: Arc<RwLock<HashMap<String, BcsInputOwner>>>,
    pub focus_changed: BcsSignal<BcsInputOwner>,
}

impl Default for BcsInputArbitrator {
    fn default() -> Self {
        Self::new()
    }
}

impl BcsInputArbitrator {
    pub fn new() -> Self {
        BcsInputArbitrator {
            active_owners: Arc::new(RwLock::new(HashMap::new())),
            focus_changed: BcsSignal::new(),
        }
    }

    // RegisterOwner adds a new network or local user to the arbitrator's pool
    pub fn register_owner(&self, owner: BcsInputOwner) {
        let mut map = self.active_owners.write().unwrap();
        map.insert(owner.id.clone(), owner);
    }

    // RequestFocus attempts to route the target focus token to the specified owner
    pub fn request_focus(&self, owner_id: &str, target_widget: Option<&Rc<RefCell<BcsObject>>>) -> bool {
        let map = self.active_owners.read().unwrap();
        let owner = map.get(owner_id);

        if owner.is_none() || target_widget.is_none() {
            return false;
        }

        // Emit signal broadcasting focus shift logic
        self.focus_changed.emit_async(owner.unwrap().clone());
        true
    }
}
